controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java

```diff
@@ -453,9 +453,15 @@ public LockedApplication storeWithUpdatedConfig(LockedApplication application, A
                 .map(Deployment::applicationVersion)
                 .collect(Collectors.toSet());
 
-        // Remove any version not deployed anywhere
-        for (ApplicationVersion version : application.get().versions()) {
-            if (!existingVersions.contains(version)) application = application.withoutVersion(version);
+        var oldVersions = application.get().versions()
+                .stream()
+                .filter(version -> !existingVersions.contains(version))
+                .sorted()
+                .collect(Collectors.toList());
+
+        // Remove any version not deployed anywhere - but keep one
+        for (int i = 0; i < oldVersions.size() - 1; i++) {
```

Comment:
Hmm. Still worried about getting holes. I was thinking of finding the oldest version deployed anywhere, and then remove everything older than this, but always keep one. That's what we do for packages we store.